Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -357,4 +360,209 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
   · rw [← hx'] at hx
     exact Or.inr <| Or.inr hx
 
+theorem two_nsmul (a : Nimber) : 2 • a = 0 := by
+  rw [succ_nsmul, one_nsmul, add_self]
+
+theorem two_zsmul (a : Nimber) : (2 : ℤ) • a = 0 :=
+  two_nsmul a
+
+theorem add_eq_iff_eq_add : a + b = c ↔ a = c + b :=
+  sub_eq_iff_eq_add
+
+theorem eq_add_iff_add_eq : a = b + c ↔ a + c = b :=
+  eq_sub_iff_add_eq
+
+/-! ### Nimber multiplication -/
+
+/-- Nimber multiplication is recursively defined so that `a * b` is the smallest nimber not equal to
+`a' * b + a * b' + a' * b'` for `a' < a` and `b' < b`. -/
+-- We write the binders like this so that the termination checker works.
+protected def mul (a b : Nimber.{u}) : Nimber.{u} :=
+  sInf {x | ∃ a', ∃ (_ : a' < a), ∃ b', ∃ (_ : b' < b),
+    Nimber.mul a' b + Nimber.mul a b' + Nimber.mul a' b' = x}ᶜ
+termination_by (a, b)
+
+instance : Mul Nimber :=
+  ⟨Nimber.mul⟩
+
+theorem mul_def (a b : Nimber) :
+    a * b = sInf {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ := by
+  change Nimber.mul a b = _
+  rw [Nimber.mul]
+  simp_rw [exists_prop]
+  rfl
+
+/-- The set in the definition of `Nimber.mul` is nonempty. -/
+private theorem mul_nonempty (a b : Nimber.{u}) :
+    {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ.Nonempty := by
+  convert nonempty_of_not_bddAbove <| not_bddAbove_compl_of_small
+    ((fun x ↦ x.1 * b + a * x.2 + x.1 * x.2) '' Set.Iio a ×ˢ Set.Iio b)
+  ext
+  simp_rw [Set.mem_setOf_eq, Set.mem_image, Set.mem_prod, Set.mem_Iio, Prod.exists]
+  tauto
+
+theorem exists_of_lt_mul (h : c < a * b) : ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = c := by
+  rw [mul_def] at h
+  have := not_mem_of_lt_csInf' h
+  rwa [Set.not_mem_compl_iff] at this
+
+theorem mul_le_of_forall_ne (h : ∀ a' < a, ∀ b' < b, a' * b + a * b' + a' * b' ≠ c) :
+    a * b ≤ c := by
+  by_contra! h'
+  have := exists_of_lt_mul h'
+  tauto
+
+instance : MulZeroClass Nimber where
+  mul_zero a := by
+    rw [← Nimber.le_zero]
+    apply mul_le_of_forall_ne
+    intro _ _ _ h
+    exact (Nimber.not_lt_zero _ h).elim
+  zero_mul a := by
+    rw [← Nimber.le_zero]
+    apply mul_le_of_forall_ne
+    intro _ h
+    exact (Nimber.not_lt_zero _ h).elim
```

Comment:
Perhaps these should also be protected theorems, to match with what's going on in the rest of the instances?

Reply:
These felt short enough to simply inline, and I (iirc) did need this instance to prove other stuff before building the whole `CommRing` instance.

Reply:
Agreed, this looks fine as is; though I'd be tempted to combine the `apply` and `intro` into a single `refine`.